ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java

```diff
@@ -488,91 +523,106 @@ private void validateState(HttpContent httpContent)
       }
     }
   }
-}
-
-/**
- * Callback for each write into the given {@link AsyncWritableChannel}.
- */
-class ContentWriteCallback implements Callback<Long> {
-  private final HttpContent httpContent;
-  private final boolean isLast;
-  private final ReadIntoCallbackWrapper callbackWrapper;
 
   /**
-   * Creates a new instance of ContentWriteCallback.
-   * @param httpContent the {@link HttpContent} whose bytes were just written. Should be null if the data from the
-   *                    original {@link HttpContent} was copied and not "retained".
-   * @param isLast if this is the last piece of {@link HttpContent} for this request.
-   * @param callbackWrapper the {@link ReadIntoCallbackWrapper} that will receive updates of bytes read and one that
-   *                        should be invoked in {@link #onCompletion(Long, Exception)} if {@code isLast} is
-   *                        {@code true} or exception passed is not null.
+   * Continues reading from the channel if the buffer watermark has not been reached and if auto-read is off.
+   * @param delta number of bytes read from the channel in the current read.
    */
-  public ContentWriteCallback(HttpContent httpContent, boolean isLast, ReadIntoCallbackWrapper callbackWrapper) {
-    this.httpContent = httpContent;
-    this.isLast = isLast;
-    this.callbackWrapper = callbackWrapper;
+  private void continueReadIfPossible(long delta) {
+    if (!channel.config().isAutoRead()) {
+      if (bytesBuffered.addAndGet(delta) < bufferWatermark) {
+        channel.read();
```

Comment:
Will fix the doc.

About `delta`, I have used the same name that is used in `AtomicLong` where this value can be negative. Do you have suggestions for names that might be clearer?